src/uniform_iterator.jl

```diff
@@ -0,0 +1,139 @@
+NOBRANCHES = Vector{Branch}()
+
+"""
+    mutable struct UniformIterator
+
+Inner iterator that enumerates all candidate programs of a uniform tree.
+- `solver`: the uniform solver.
+- `outeriter`: outer iterator that is responsible for producing uniform trees. This field is used to dispatch on the [`derivation_heuristic`](@ref).
+- `unvisited_branches`: for each search-node from the root to the current search-node, a list of unviisted branches.
+- `nsolutions`: number of solutions found so far.
+"""
+mutable struct UniformIterator
+    solver::UniformSolver
+    outeriter::Union{ProgramIterator, Nothing}
+    unvisited_branches::Stack{Vector{Branch}}
+    nsolutions::Int
+end
+
+"""
+    UniformIterator(solver::UniformSolver, outeriter::ProgramIterator)
+
+Constructs a new UniformIterator that traverses solutions of the [`UniformSolver`](@ref) and is an inner iterator of an outer [`ProgramIterator`](@ref).
+"""
+function UniformIterator(solver::UniformSolver, outeriter::Union{ProgramIterator, Nothing})
+    iter = UniformIterator(solver, outeriter, Stack{Vector{Branch}}(), 0)
+    if isfeasible(solver)
+        # create search-branches for the root search-node
+        save_state!(solver)
+        push!(iter.unvisited_branches, generate_branches(iter))
+    end
+    return iter
+end
+
+"""
+Returns a vector of disjoint branches to expand the search tree at its current state.
+Example:
+```
+# pseudo code
+Hole(domain=[2, 4, 5], children=[
+    Hole(domain=[1, 6]), 
+    Hole(domain=[1, 6])
+])
+```
+If we split on the first hole, this function will create three branches.
+- `(firsthole, 2)`
+- `(firsthole, 4)`
+- `(firsthole, 5)`
+"""
+function generate_branches(iter::UniformIterator)::Vector{Branch}
+    @assert isfeasible(iter.solver)
+    function _dfs(node::Union{StateHole, RuleNode})
+        if node isa StateHole && size(node.domain) > 1
+            #skip the derivation_heuristic if the parent_iterator is not set up
+            if isnothing(iter.outeriter)
+                return [(node, rule) for rule ∈ node.domain]
+            end
+            #reversing is needed because we pop and consider the rightmost branch first
+            return reverse!([(node, rule) for rule ∈ derivation_heuristic(iter.outeriter, findall(node.domain))])
+        end
+        for child ∈ node.children
+            branches = _dfs(child)
+            if !isempty(branches)
+                return branches
+            end
+        end
+        return NOBRANCHES
+    end
+    return _dfs(get_tree(iter.solver))
+end
+
+"""
+    next_solution!(iter::UniformIterator)::Union{RuleNode, StateHole, Nothing}
+
+Searches for the next unvisited solution.
+Returns nothing if all solutions have been found already.
+"""
+function next_solution!(iter::UniformIterator)::Union{RuleNode, StateHole, Nothing}
+    solver = iter.solver
+    if iter.nsolutions == 1000000 @warn "UniformSolver is iterating over more than 1000000 solutions..." end
```

Comment:
I thinks this check is probably for debugging purposes. Should it be included in the actual implementation?

Reply:
We discussed this and wanted to keep it for now. Good eye, though—thank you for mentioning it!